R/tag_query.R

```diff
@@ -0,0 +1,1601 @@
+          self
+        },
+        ## end Adjust Children
+
+        #' ## Adjust Siblings
+        #'
+        #' * `$after(...)`: Add all `...` objects as siblings after each of the
+        #' selected elements.
+        after = function(...) {
+          rebuild_()
+          tagQuerySiblingAfter(selected_, ...)
+          rebuild_()
+          self
+        },
+        #' * `$before(...)`: Add all `...` objects as siblings before each of
+        #' the selected elements.
+        before = function(...) {
+          rebuild_()
+          tagQuerySiblingBefore(selected_, ...)
+          rebuild_()
+          self
+        },
+        #' * `$replaceWith(...)`: Replace all selected elements with `...`. This
+        #' also sets the selected elements to an empty set. A new `tagQuery()`
+        #' object will be created with an empty set of selected elements.
+        replaceWith = function(...) {
+          rebuild_()
+          tagQuerySiblingReplaceWith(selected_, ...)
+          newTagQuery(list(), rebuild = TRUE)
+        },
+        #' * `$remove(...)`: Remove all selected elements from the `tagQuery()`
+        #' object. The selected elements is set to an empty set. A new
+        #' `tagQuery()` object will be created with an empty set of selected
+        #' elements.
+        remove = function() {
+          rebuild_()
+          tagQuerySiblingRemove(selected_)
+          # Remove items from selected info
+          newTagQuery(list(), rebuild = TRUE)
+        },
+        ## end Adjust Siblings
+
+        ## Generic Methods
+        #' ## Generic methods
+        #'
+        #' * `$each(fn)`: Perform function `fn` on each of the selected
+        #' elements. `fn` should accept two arguments: a selected element and
+        #' the selected element's position within the selected elements. This
+        #' argument order is different than jQuery's `$().each()` as there is no
+        #' concept of a `this` object inside the function execution. To stay
+        #' consistent with other methods, the each of the selected tag
+        #' environments will be given first, followed by the index position. Any
+        #' alterations to the provided tag environments will persist in calling
+        #' tag query object.
+        each = function(fn) {
+          rebuild_()
+          tagQueryEach(selected_, fn)
+          rebuild_()
+          self
+        },
+        ## end Generic Methods
+
+        #' ## Tag Query methods
+        #'
+        #' * `$asTags(selected = TRUE)`: If `selected = TRUE`, then all
+        #' previously found elements (and their descendants) will be
+        #' converted to tags. If `selected = FALSE`, the top level tag
+        #' elements (and their descendants) will be converted to
+        #' standard tags. If there is more than one element being
+        #' returned, a `tagList()` will be used to hold all of the
+        #' objects.
+        asTags = function(selected = TRUE) {
```

Comment:
It's unfortunate that the return type of this isn't stable, but I can't think of a better way to go about it.

Reply:
We could always return a `tagList()`, even if there is a single element. 

`flattenTagsRaw()` will make sure it doesn't ever wrap `tagList()`s on `tagList()`s.